homeassistant/components/saj/sensor.py

```diff
@@ -48,139 +58,161 @@
+    await inverter.setup(hass, async_add_entities)
 
-    remove_interval_update = None
-    wifi = config[CONF_TYPE] == INVERTER_TYPES[1]
-
-    # Init all sensors
-    sensor_def = pysaj.Sensors(wifi)
-
-    # Use all sensors by default
-    hass_sensors = []
-
-    kwargs = {}
-    if wifi:
-        kwargs["wifi"] = True
-        if config.get(CONF_USERNAME) and config.get(CONF_PASSWORD):
-            kwargs["username"] = config[CONF_USERNAME]
-            kwargs["password"] = config[CONF_PASSWORD]
-
-    try:
-        saj = pysaj.SAJ(config[CONF_HOST], **kwargs)
-        done = await saj.read(sensor_def)
-    except pysaj.UnauthorizedException:
-        _LOGGER.error("Username and/or password is wrong")
-        return
-    except pysaj.UnexpectedResponseException as err:
-        _LOGGER.error(
-            "Error in SAJ, please check host/ip address. Original error: %s", err
-        )
-        return
 
-    if not done:
-        raise PlatformNotReady
+async def async_setup_platform(
+    hass: HomeAssistant, config, async_add_entities: Callable, discovery_info=None
+):  # pragma: no cover
+    """Set up the SAJ inverter with yaml."""
+    _LOGGER.warning(
+        "Loading SAJ Solar inverter integration via yaml is deprecated. "
+        "Please remove it from your configuration"
+    )
+    inverter = SAJInverter(config)
+    await inverter.setup(hass, async_add_entities)
 
-    for sensor in sensor_def:
-        if sensor.enabled:
-            hass_sensors.append(
-                SAJsensor(saj.serialnumber, sensor, inverter_name=config.get(CONF_NAME))
-            )
 
-    async_add_entities(hass_sensors)
+def _init_pysaj(wifi, config):  # pragma: no cover
+    kwargs = {"wifi": wifi}
+    if config.get(CONF_USERNAME) and config.get(CONF_PASSWORD):
+        kwargs["username"] = config[CONF_USERNAME]
+        kwargs["password"] = config[CONF_PASSWORD]
 
-    async def async_saj():
-        """Update all the SAJ sensors."""
-        values = await saj.read(sensor_def)
+    return pysaj.SAJ(config[CONF_HOST], **kwargs)
 
-        for sensor in hass_sensors:
-            state_unknown = False
-            # SAJ inverters are powered by DC via solar panels and thus are
-            # offline after the sun has set. If a sensor resets on a daily
-            # basis like "today_yield", this reset won't happen automatically.
-            # Code below checks if today > day when sensor was last updated
-            # and if so: set state to None.
-            # Sensors with live values like "temperature" or "current_power"
-            # will also be reset to None.
-            if not values and (
-                (sensor.per_day_basis and date.today() > sensor.date_updated)
-                or (not sensor.per_day_basis and not sensor.per_total_basis)
-            ):
-                state_unknown = True
-            sensor.async_update_values(unknown_state=state_unknown)
 
-        return values
+class SAJInverter:
```

Comment:
This is not part of the sensor anymore, I suggest moving this out of the sensor file.

This class could be extending the DataUpdateCoordinator (instead of wrapping it).